docs: add postgres-db guide (#719)
nitrictech/docs: docs/guides/nodejs/nitric-and-pgsql.mdx

---
description: 'How to work with Nitric SQL Databases using PostgreSQL'
tags:
  - Databases & CMS
languages:
  - typescript
  - javascript

> published_at: 2025-03-18

Reviewer: ```suggestion
published_at: 2025-04-03
```

---

# Nitric SQL Databases with PostgreSQL

This guide demonstrates how to use PostgreSQL with Nitric's [SQL database](/sql) feature. We will be creating a simple to-do app. The finished source can be found [here](https://github.com/nitrictech/examples/tree/main/v1/nitric-pgsql).

<Note>SQL databases are currently in preview and only support PostgreSQL.</Note>

## Create a new Nitric project

The first step is to create a new Nitric TypeScript project using the [Nitric CLI](/reference/cli).

```bash
nitric new todo-app ts-starter
```

## Enabling SQL databases

SQL databases are currently in Preview. To enable this feature in your project, add the following to your `nitric.yaml` file:

```yaml title: nitric.yaml
preview:
  - sql-databases
```

## Add a PostgreSQL client

In this guide, we'll use the [node-postgres](https://github.com/brianc/node-postgres) library, but feel free to use another client of your choice and adjust the steps accordingly.

```bash
npm install pg
```

## Create a Nitric SQL Database and Connect to it

> Use Nitric’s `sql` feature to create a new SQL database, then use connection string and the `pg` library to connect to it.

Reviewer 2: ```suggestion
Use Nitric’s `sql` feature to create a new SQL database, then use `.connectionString()` and the `pg` library to connect to it.
```

```ts title: resources/db.ts
import { sql } from '@nitric/sdk'
import { Client } from 'pg'

const db = sql('todo-db', {
  migrations: 'file://migrations/todos',
})

let client: Client

const getClient = async () => {
  if (!client) {
    const connectionString = await db.connectionString()
    client = new Client({ connectionString })
    await client.connect()
  }
  return client
}

export default getClient
```

## Database Migrations

Create a migration file to define the `todos` table.

```sql title: migrations/todos/1_create_table.up.sql
CREATE TABLE IF NOT EXISTS todos (
    id SERIAL PRIMARY KEY,

>     text TEXT NOT NULL,

Reviewer 2: There are highlighting issues due to "text" being a keyword in postgres... might be worth changing to "content" if updating all the screenshots isn't a massive pain. Otherwise, I think its fine.

Author: Updated screenshots and code.

---
description: 'How to work with Nitric SQL Databases using PostgreSQL'
tags:
  - Databases & CMS
languages:
  - typescript
  - javascript
published_at: 2025-03-18
---

# Nitric SQL Databases with PostgreSQL

This guide demonstrates how to use PostgreSQL with Nitric's [SQL database](/sql) feature. We will be creating a simple to-do app. The finished source can be found [here](https://github.com/nitrictech/examples/tree/main/v1/nitric-pgsql).

<Note>SQL databases are currently in preview and only support PostgreSQL.</Note>

## Create a new Nitric project

The first step is to create a new Nitric TypeScript project using the [Nitric CLI](/reference/cli).

```bash
nitric new todo-app ts-starter
```

## Enabling SQL databases

SQL databases are currently in Preview. To enable this feature in your project, add the following to your `nitric.yaml` file:

```yaml title: nitric.yaml
preview:
  - sql-databases
```

> ## Install PostgreSQL client

Reviewer 3: ```suggestion
## Add a PostgreSQL client
```

> 

Reviewer 3: ```suggestion

In this guide, we'll use the [node-postgres](https://github.com/brianc/node-postgres) library, but feel free to use another client of your choice and adjust the steps accordingly.

```

```bash
npm install pg
```

## Create a Nitric SQL Database and Connect to it

> Establish a connection to a PostgreSQL database using Nitric’s `sql` feature and the `pg` library.

Reviewer 3: ```suggestion
Use Nitric’s `sql` feature to create a new SQL database, then use connection string and the `pg` library to connect to it.
```

```ts title: resources/db.ts
import { sql } from '@nitric/sdk'
import { Client } from 'pg'

const db = sql('todo-db', {
  migrations: 'file://migrations/todos',
})

let client: Client

const getClient = async () => {
  if (!client) {
    const connectionString = await db.connectionString()
    client = new Client({ connectionString })
    await client.connect()
  }
  return client
}

export default getClient
```

## Database Migrations

Create a migration file to define the `todos` table.

```sql title: migrations/todos/1_create_table.up.sql
CREATE TABLE IF NOT EXISTS todos (
    id SERIAL PRIMARY KEY,
    text TEXT NOT NULL,
    done BOOLEAN DEFAULT FALSE
);
```

## Developing the API

### Building the API

Delete the `services/hello.ts` file and create a new one called `todos.ts`, which will contain our API endpoints.

```ts title: services/todos.ts
import { api } from '@nitric/sdk'
import getClient from '../resources/db'

const mainApi = api('main')

// Fetch all todos
mainApi.get('/todos', async (ctx) => {
  const client = await getClient()
  const result = await client.query('SELECT * FROM todos ORDER BY id ASC')
  return ctx.res.json(result.rows)
})

> // Insert a new todo
> mainApi.post('/todos/:id', async (ctx) => {
>   const { id } = ctx.req.params
>   const { text } = ctx.req.json()
>   const client = await getClient()
>   await client.query('INSERT INTO todos (id, text, done) VALUES ($1, $2, $3)', [
>     parseInt(id),
>     text,
>     false,
>   ])
> })

Reviewer 3: ```suggestion
// Insert a new todo
mainApi.post('/todos/', async (ctx) => {
  const { text } = ctx.req.json()
  const client = await getClient()
  await client.query('INSERT INTO todos (text, done) VALUES ($1, $2)', [
    text,
    false,
  ])
})
```

The id was already a serial (autoincrement), it shouldn't be inserted here.

// Update text of a todo
mainApi.patch('/todos/:id', async (ctx) => {
  const { id } = ctx.req.params
  const { text } = ctx.req.json()
  const client = await getClient()
  await client.query('UPDATE todos SET text = $1 WHERE id = $2', [
    text,
    parseInt(id),
  ])
})

// Toggle todo completion status

> mainApi.patch('/todos/:id/toggle', async (ctx) => {
>   const { id } = ctx.req.params
>   const client = await getClient()
>   const todo = await client.query('SELECT done FROM todos WHERE id = $1', [
>     parseInt(id),
>   ])
> 
>   if (todo.rows.length > 0) {
>     const newDone = !todo.rows[0].done
>     await client.query('UPDATE todos SET done = $1 WHERE id = $2', [
>       newDone,
>       parseInt(id),
>     ])
>   }
> })

Reviewer 3: ```suggestion
mainApi.patch('/todos/:id/toggle', async (ctx) => {
  const { id } = ctx.req.params
  const client = await getClient()
  
  await client.query('UPDATE todos SET done = NOT done WHERE id = $1',
    [parseInt(id)]
  );
})
```

No need to query twice

// Delete a todo
mainApi.delete('/todos/:id', async (ctx) => {
  const { id } = ctx.req.params
  const client = await getClient()
  await client.query('DELETE FROM todos WHERE id = $1', [parseInt(id)])
})
```

### Start testing with `nitric start`

Run `nitric start` to start your local database and APIs.

```bash
nitric start
```

Using the dashboard, apply your migrations from the databases tab to initialize your local db.

![Apply migrations](/docs/images/guides/nitric-and-pgsql/migrations.png)

### Add some todos using the Nitric dashboard

> Open the local dashboard at <a target="_blank" href="http://localhost:49152">localhost:49152</a>, then navigate to the `POST /todos/{id}` endpoint and fill in an `id` as the path param.

Reviewer 3: ```suggestion
Open the local dashboard at <a target="_blank" href="http://localhost:49152">localhost:49152</a>, then navigate to the `POST /todos` endpoint.
```

> ![Add id param for creating a todo](/docs/images/guides/nitric-and-pgsql/step-1.png)

Reviewer 3: This needs to be updated with a new screenshot, the ID shouldn't be provided when the todo item is first created, it's autogenerated

---
description: 'How to work with Nitric SQL Databases using PostgreSQL'
tags:
  - Databases & CMS
languages:
  - typescript
  - javascript
published_at: 2025-03-18
---

# Nitric SQL Databases with PostgreSQL

This guide demonstrates how to use PostgreSQL with Nitric's [SQL database](/sql) feature. We will be creating a simple to-do app. The finished source can be found [here](https://github.com/nitrictech/examples/tree/main/v1/nitric-pgsql).

<Note>
  SQL databases are currently in preview and only support PostgreSQL.
</Note>

## Create a new Nitric project

The first step is to create a new Nitric TypeScript project using the [Nitric CLI](/reference/cli).

```bash
nitric new todo-app ts-starter

> npm install

Reviewer: ```suggestion
```

Reviewer: Maybe just make a note to install dependencies or cd into it.

```

## Enabling SQL databases

SQL databases are currently in Preview. To enable this feature in your project, add the following to your `nitric.yaml` file:

```yaml title: nitric.yaml
preview:
  - sql-databases
```

> Ensure your CLI is at least `v1.42` and your AWS provider is at least `v1.6.2`.

Reviewer: We support sql on all clouds on pulumi now so lets update it to mention we support all but will choose AWS for this guide.

## Install PostgreSQL client

```bash
npm install pg
```

## Create a Nitric SQL Database and Connect to it

Establish a connection to a PostgreSQL database using Nitric’s `sql` feature and the `pg` library.

```ts title: resources/db.ts
import { sql } from '@nitric/sdk'
import { Client } from 'pg'

const db = sql('todo-db', {
  migrations: 'file://migrations/todos',
})

let client: Client

const getClient = async () => {
  if (!client) {
    const connectionString = await db.connectionString()
    client = new Client({ connectionString })
    await client.connect()
  }
  return client
}

export default getClient
```

## Database Migrations

Create a migration file to define the `todos` table.

> ```sql title: 'migrations/1_create_table.up.sql'

Reviewer: ```suggestion
```sql title: migrations/todos/1_create_table.up.sql
```

---
description: 'How to work with Nitric SQL Databases using PostgreSQL'
tags:

>   - Databases

Reviewer 2: ```suggestion
  - Databases & CMS
```

languages:
  - typescript
  - javascript
published_at: 2025-03-18
---

# Nitric SQL Databases with PostgreSQL

This guide demonstrates how to use PostgreSQL with Nitric's [SQL database](/sql) feature. We will be creating a simple to-do app. The finished source can be found [here](https://github.com/nitrictech/examples/tree/main/v1/nitric-pgsql).

<Note>

>   SQL databases are currently in preview and only support PostgreSQL deployed to
>   AWS.

Reviewer 2: ```suggestion
  SQL databases are currently in preview and only support PostgreSQL.
```
